src/main/java/edu/kit/provideq/toolbox/featuremodel/anomaly/dead/SatBasedDeadFeatureSolver.java

```diff
@@ -1,115 +1,116 @@
+    return checkDeadFeatures(cnf, subRoutineResolver);
   }
 
-  private static void checkDeadFeatures(Solution<String> solution, String cnf,
-                                        Function<String, Solution<DimacsCnfSolution>> satSolve) {
+  @Override
+  public ProblemType<String, String> getProblemType() {
+    return DeadFeatureConfiguration.FEATURE_MODEL_ANOMALY_DEAD;
+  }
+
+  private static Mono<Solution<String>> checkDeadFeatures(
+      String cnf,
+      SubRoutineResolver subRoutineResolver
+  ) {
     // Check if there are any Dead Features
     DimacsCnf dimacsCnf;
     try {
       dimacsCnf = DimacsCnf.fromDimacsCnfString(cnf);
     } catch (ConversionException e) {
+      var solution = new Solution<String>();
       solution.setDebugData("Conversion error: " + e.getMessage());
       solution.abort();
-      return;
-    }
-
-    var builder = new StringBuilder();
-    var errorBuilder = new StringBuilder();
-
-    for (Variable variable : dimacsCnf.getVariables()) {
-      // Use formula: ¬SAT (FM ∧ f) to check for a dead feature
-      // So add variable to the cnf of the feature model and check if there is a solution
-      // If there is a solution, the feature is not dead
-      var orClause = new ArrayList<Variable>();
-      orClause.add(variable);
-      var variableCnf = dimacsCnf.addOrClause(orClause);
-
-      var variableSolution = satSolve.apply(variableCnf.toString());
-
-      if (variableSolution.getStatus() == SolutionStatus.SOLVED) {
-        var dimacsCnfSolution =
-            DimacsCnfSolution.fromString(dimacsCnf, variableSolution.getSolutionData().toString());
-
-        if (dimacsCnfSolution.isVoid()) {
-          builder.append(variable.name())
-              .append("\n");
-        }
-      } else {
-        errorBuilder.append(variableSolution.getDebugData())
-            .append("\n");
-      }
-    }
-
-    if (builder.isEmpty()) {
-      builder.append("No features are dead features!\n");
-    } else {
-      builder.insert(0, "The following features are dead features:\n");
-    }
-
-    if (!errorBuilder.isEmpty()) {
-      builder.append("Following errors occurred:\n")
-          .append(errorBuilder);
-
-      solution.fail();
+      return Mono.just(solution);
     }
 
-    solution.setSolutionData(builder.toString());
-    solution.complete();
+    return Flux.fromIterable(dimacsCnf.getVariables())
+        .flatMap(variable -> {
+          // Use formula: ¬SAT (FM ∧ f) to check for a dead feature
+          // So add variable to the cnf of the feature model and check if there is a solution
+          // If there is a solution, the feature is not dead
+
+          var variableCheckCnf = dimacsCnf.addOrClause(new ArrayList<>(List.of(variable)));
+          return subRoutineResolver
+              .runSubRoutine(SAT_SUBROUTINE, variableCheckCnf.toString())
+              .map(variableCheckSolution -> DimacsCnfSolution.fromString(dimacsCnf, variableCheckSolution.getSolutionData().toString()))
```

Comment:
Checkstyle: Can you split this to multiple lines?

Reply:
Fixed in 8fa5049. I also improved the readability of the code in general

Reply:
awesome, thanks a lot!